include/fiction/technology/sidb_cluster_hierarchy.hpp

```diff
@@ -0,0 +1,1083 @@
+        }
+        else if constexpr (bound == bound_direction::UPPER)
+        {
+            return *std::prev(std::find_if(order.crbegin(), order.crend(),
+                                           [&](const potential_projection& pp) { return pp.multiset == m_conf; })
+                                  .base(),
+                              1);
+        }
+    }
+    /**
+     * Removes all occurrences of potential projections that match their multiset charge configuration to the argument.
+     *
+     * @param m_conf The multiset charge configuration of which all occurrences must be removed.
+     */
+    void remove_m_conf(const uint64_t m_conf) noexcept
+    {
+        for (pot_proj_order::const_iterator it = order.cbegin(); it != order.cend();)
+        {
+            it->multiset == m_conf ? it = order.erase(it) : ++it;
+        }
+    }
+    /**
+     * Adds a potential projection to the ordering, automatically placing it appropriately.
+     *
+     * @param pp Potential projection to add.
+     */
+    inline void add(const potential_projection& pp) noexcept
+    {
+        order.emplace(pp);
+    }
+};
+/**
+ * The type of a charge space is defined. It is a set of cluster charge states, forming the state space in our setting.
+ */
+#ifdef DEBUG_SIDB_CLUSTER_HIERARCHY
+using sidb_cluster_charge_state_space = std::unordered_set<sidb_cluster_charge_state, sidb_cluster_charge_state>;
+#else
+using sidb_cluster_charge_state_space = phmap::flat_hash_set<sidb_cluster_charge_state, sidb_cluster_charge_state>;
+#endif
+/**
+ * This struct defines a hashing of shared pointers to clusters, used to facilitate quick access in a clustering.
+ */
+struct sidb_cluster_ptr_hash
+{
+    /**
+     * The hashing operation is defined.
+     *
+     * @param c Shared pointer to a cluster to take the has of.
+     * @return The hash computed over the the unique id associated with the cluster.
+     */
+    inline std::size_t operator()(const sidb_cluster_ptr& c) const noexcept
+    {
+        return std::hash<uint64_t>{}(get_unique_cluster_id(c));
+    }
+};
+/**
+ * A clustering is a set of disjoint clusters, ie., none share an SiDB.
+ */
+#ifdef DEBUG_SIDB_CLUSTER_HIERARCHY
+using sidb_clustering = std::set<sidb_cluster_ptr>;
+#else
+using sidb_clustering = phmap::flat_hash_set<sidb_cluster_ptr, sidb_cluster_ptr_hash>;
+#endif
+/**
+ * This struct defined the fully decorated cluster hierarchy type which follows the structure of a "general tree". It
+ * contains multiple stores:
+ * - a set of indices that correspond to the SiDBs as stored in a `charge_distribution_surface` object,
+ * - a store of potential bounds for each SiDB in the cluster that give the fraction of the local electrostatic
+ * potential that is received from outside the cluster, and
+ * - a store containing the potential projection orders onto each SiDB, inside and outside the cluster.
+ */
+struct sidb_cluster
+{
+    using uid_t   = uint64_t;
```

Comment:
We don't do that in _fiction_ :)

Reply:
No typedefs in structs? Or no typedefs of base types? In any case I would argue that semantical difference that is presented with these typedefs between `uid`s and `sidb_ix`s only helps the code clarity, and they are contained to the struct.

Reply:
In General, we don't use typedefs of base types. I see your point. We can also leave it for now.